### PR TITLE
Add Sitemap support

### LINKED ISSUE #6: Add support for Sitemap plugin
This plugin should register a custom item types to the Page plugin to be able to add to the Sitemap plugin at least all channels urls and all topics urls.

Currently toying with it to build a tiny forum, I'll probably make a PR soon, if any suggestion on what should be or not.

### PR DESCRIPTION
**Pull request by RomainMazB**
@bennothommo I messed up with the previous PR while reverting the versions file commit.

Ready for reviews (same modification as previous except the version file).

### CODE REVIEW

File: Plugin.php [hunk 1]
```diff
@@ -91,6 +93,30 @@ public function boot()
                 ]
             ]);
         });
+
+        Event::listen('pages.menuitem.listTypes', function() {
```
Comment (bennothommo) [inline Plugin.php:97]:
```suggestion
        Event::listen('pages.menuitem.listTypes', function () {
```

File: Plugin.php [hunk 2]
```diff
@@ -91,6 +93,30 @@ public function boot()
...
+            return [
+                'forum-channel' => 'winter.forum::lang.menuitem.forum_channel',
+                'all-forum-channels' => 'winter.forum::lang.menuitem.all_forum_channels',
+                'all-forum-topics' => 'winter.forum::lang.menuitem.all_forum_topics',
+            ];
+        });
+
+        Event::listen('pages.menuitem.getTypeInfo', function($type) {
```
Comment (bennothommo) [inline Plugin.php:105]:
```suggestion
        Event::listen('pages.menuitem.getTypeInfo', function ($type) {
```

File: Plugin.php [hunk 3]
```diff
@@ -91,6 +93,30 @@ public function boot()
...
+            if ($type == 'forum-channel' || $type == 'all-forum-channels') {
+                return Channel::getMenuTypeInfo($type);
+            } elseif ($type == 'forum-topic' || $type == 'all-forum-topics') {
+                return Topic::getMenuTypeInfo($type);
+            }
```
Comment (bennothommo) [inline Plugin.php:110]: @RomainMazB as a rule, we should use strict comparisons (`===`) in most cases. It allows us to uncover subtle bugs a lot easier.

Also, I didn't see `forum-topic` as an applicable option in the `pages.menuitem.listTypes` event?

```suggestion
            if ($type === 'forum-channel' || $type === 'all-forum-channels') {
                return Channel::getMenuTypeInfo($type);
            } elseif ($type === 'all-forum-topics') {
                return Topic::getMenuTypeInfo($type);
            }
```

File: Plugin.php [hunk 4]
```diff
@@ -91,6 +93,30 @@ public function boot()
...
+        });
+
+        Event::listen('pages.menuitem.resolveItem', function($type, $item, $url, $theme) {
+            if ($type == 'forum-channel' || $type == 'all-forum-channels') {
+                return Channel::resolveMenuItem($item, $url, $theme);
+            } elseif ($type == 'all-forum-topics') {
+                return Topic::resolveMenuItem($item, $url, $theme);
+            }
+        });
```
Comment (bennothommo) [inline Plugin.php:119]:
```suggestion
        Event::listen('pages.menuitem.resolveItem', function ($type, $item, $url, $theme) {
            if ($type === 'forum-channel' || $type === 'all-forum-channels') {
                return Channel::resolveMenuItem($item, $url, $theme);
            } elseif ($type === 'all-forum-topics') {
                return Topic::resolveMenuItem($item, $url, $theme);
            }
        });
```

File: models/Channel.php [hunk 5]
```diff
@@ -2,6 +2,9 @@
 
 use Model;
 use ApplicationException;
+use Url;
+use Cms\Classes\Page as CmsPage;
+use Cms\Classes\Theme;
```
Comment (bennothommo) [inline models/Channel.php:7]:
```suggestion
use Url;
use Model;
use ApplicationException;
use Cms\Classes\Page as CmsPage;
use Cms\Classes\Theme;
```

File: models/Channel.php [hunk 6]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
      */
+    public static function getMenuTypeInfo($type)
+    {
+        $result = [];
+
+        if ($type == 'forum-channel') {
+            $result = [
+                'references' => self::listChildrenOptions(),
+                'nesting' => true,
+                'dynamicItems' => true
```
Comment (bennothommo) [inline models/Channel.php:108]: @RomainMazB where possible, please ensure that the last option in an array has a trailing comma. It helps with file diffs.
```suggestion
                'dynamicItems' => true,
```

File: models/Channel.php [hunk 7]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+     * - cmsPages - a list of CMS pages (objects of the Cms\Classes\Page class), if the item type requires a CMS page reference to
+     *   resolve the item URL.
+     * @param string $type Specifies the menu item type
+     * @return array Returns an array
      */
+    public static function getMenuTypeInfo($type)
+    {
+        $result = [];
+
+        if ($type == 'forum-channel') {
```
Comment (bennothommo) [inline models/Channel.php:104]:
```suggestion
        if ($type === 'forum-channel') {
```

File: models/Channel.php [hunk 8]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+                'nesting' => true,
+                'dynamicItems' => true
+            ];
+        }
+
+        if ($type == 'all-forum-channels') {
+            $result = [
+                'dynamicItems' => true
+            ];
+        }
```
Comment (bennothommo) [inline models/Channel.php:116]:
```suggestion
        if ($type === 'all-forum-channels') {
            $result = [
                'dynamicItems' => true,
            ];
        }
```

File: models/Channel.php [hunk 9]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+     * identified by the child id as key and with the following elements for each key:
+     *  - title - The channel title
+     *  - items - An array containing the channel children (recursive)
+     * @return array
+     */
+    protected static function listChildrenOptions(): array
+    {
+        $channel = self::isVisible()->getNested();
+
+        $iterator = function($children) use (&$iterator) {
```
Comment (bennothommo) [inline models/Channel.php:157]:
```suggestion
        $iterator = function ($children) use (&$iterator) {
```

File: models/Channel.php [hunk 10]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+            $result = [];
+
+            foreach ($children as $child) {
+                if (!$child->children) {
+                    $result[$child->id] = $child->title;
+                }
+                else {
```
Comment (bennothommo) [inline models/Channel.php:164]:
```suggestion
                } else {
```

File: models/Channel.php [hunk 11]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+                    $result[$child->id] = [
+                        'title' => $child->title,
+                        'items' => $iterator($child->children)
```
Comment (bennothommo) [inline models/Channel.php:167]:
```suggestion
                        'items' => $iterator($child->children),
```

File: models/Channel.php [hunk 12]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+     * @param \Cms\Classes\Theme $theme Specifies the current theme.
+     * @param string $url Specifies the current page URL, normalized, in lower case
+     * The URL is specified relative to the website root, it includes the subdirectory name, if any.
+     * @return mixed Returns an array. Returns null if the item cannot be resolved.
+     */
+    public static function resolveMenuItem($item, $url, $theme)
+    {
+        $result = null;
+
+        if ($item->type == 'forum-channel') {
```
Comment (bennothommo) [inline models/Channel.php:199]:
```suggestion
        if ($item->type === 'forum-channel') {
```

File: models/Channel.php [hunk 13]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+            $pageUrl = self::getChannelPageUrl($item->cmsPage, $channel, $theme);
+            if (!$pageUrl) {
+                return;
+            }
+
+            $pageUrl = Url::to($pageUrl);
+
+            $result = [];
+            $result['url'] = $pageUrl;
+            $result['isActive'] = $pageUrl == $url;
```
Comment (bennothommo) [inline models/Channel.php:218]:
```suggestion
            $result['isActive'] = ($pageUrl === $url);
```

File: models/Channel.php [hunk 14]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+            $result['mtime'] = $channel->updated_at;
+
+            if ($item->nesting) {
+                $channels = $channel->isVisible()->getNested();
+                $iterator = function($channels) use (&$iterator, &$item, &$theme, $url) {
```
Comment (bennothommo) [inline models/Channel.php:223]:
```suggestion
                $iterator = function ($channels) use (&$iterator, &$item, &$theme, $url) {
```

File: models/Channel.php [hunk 15]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+
+                    return $branch;
+                };
+
+                $result['items'] = $iterator($channels);
+            }
+        }
+        elseif ($item->type == 'all-forum-channels') {
+            $result = [
+                'items' => []
```
Comment (bennothommo) [inline models/Channel.php:248]:
```suggestion
                'items' => [],
```

File: models/Channel.php [hunk 16]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+                        $branch[] = $branchItem;
+                    }
+
+                    return $branch;
+                };
+
+                $result['items'] = $iterator($channels);
+            }
+        }
+        elseif ($item->type == 'all-forum-channels') {
```
Comment (bennothommo) [inline models/Channel.php:246]:
```suggestion
        elseif ($item->type === 'all-forum-channels') {
```

File: models/Channel.php [hunk 17]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+            $result = [
+                'items' => []
+            ];
+
+            $channels = self::isVisible()->orderBy('title')->get();
+            foreach ($channels as $channel) {
+                $channelItem = [
+                    'title' => $channel->name,
+                    'url'   => self::getChannelPageUrl($item->cmsPage, $channel, $theme),
+                    'mtime' => $channel->updated_at
```
Comment (bennothommo) [inline models/Channel.php:256]:
```suggestion
                    'mtime' => $channel->updated_at,
```

File: models/Channel.php [hunk 18]
```diff
@@ -78,8 +81,226 @@ class Channel extends Model
...
+                ];
+
+                $channelItem['isActive'] = $channelItem['url'] == $url;
```
Comment (bennothommo) [inline models/Channel.php:259]:
```suggestion
                $channelItem['isActive'] = ($channelItem['url'] === $url);
```

File: models/Topic.php [hunk 19]
```diff
@@ -318,4 +321,128 @@ public function lockTopic()
...
+     *   Optional, false if omitted.
+     * - cmsPages - a list of CMS pages (objects of the Cms\Classes\Page class), if the item type requires a CMS page reference to
+     *   resolve the item URL.
+     * @param string $type Specifies the menu item type
+     * @return array Returns an array
+     */
+    public static function getMenuTypeInfo($type)
+    {
+        $result = [
+            'dynamicItems' => true
```
Comment (bennothommo) [inline models/Topic.php:345]:
```suggestion
            'dynamicItems' => true,
```

File: models/Topic.php [hunk 20]
```diff
@@ -318,4 +321,128 @@ public function lockTopic()
...
+     * @param \Winter\Pages\Classes\MenuItem $item Specifies the menu item.
+     * @param \Cms\Classes\Theme $theme Specifies the current theme.
+     * @param string $url Specifies the current page URL, normalized, in lower case
+     * The URL is specified relative to the website root, it includes the subdirectory name, if any.
+     * @return mixed Returns an array. Returns null if the item cannot be resolved.
+     */
+    public static function resolveMenuItem($item, $url, $theme)
+    {
+        $result = [
+            'items' => []
```
Comment (bennothommo) [inline models/Topic.php:394]:
```suggestion
            'items' => [],
```

File: models/Topic.php [hunk 21]
```diff
@@ -318,4 +321,128 @@ public function lockTopic()
...
+
+        $topics = self::whereHas('channel', function ($query) {
+            $query->isVisible();
+        })->orderBy('subject')->get();
+
+        foreach ($topics as $topic) {
+            $topicItem = [
+                'title' => $topic->subject,
+                'url'   => self::getTopicPageUrl($item->cmsPage, $topic, $theme),
+                'mtime' => $topic->updated_at
```
Comment (bennothommo) [inline models/Topic.php:405]:
```suggestion
                'mtime' => $topic->updated_at,
```

**Review by bennothommo — CHANGES_REQUESTED** [review]
@RomainMazB could you make the following changes and re-test for us?

File: Plugin.php (hunk 3 above)
Comment (RomainMazB) [reply Plugin.php:110]: You right, nice catch!
I first wanted to add it but then I said to myself that I don't see any use case where we could add only one topic in the sitemap.

File: models/Topic.php [hunk 22]
```diff
@@ -318,4 +321,128 @@ public function lockTopic()
...
+        })->orderBy('subject')->get();
+
+        foreach ($topics as $topic) {
+            $topicItem = [
+                'title' => $topic->subject,
+                'url'   => self::getTopicPageUrl($item->cmsPage, $topic, $theme),
+                'mtime' => $topic->updated_at,
+            ];
+
+            $topicItem['isActive'] = $topicItem['url'] == $url;
```
Comment (bennothommo) [inline models/Topic.php:408]:
```suggestion
            $topicItem['isActive'] = ($topicItem['url'] === $url);
```

File: models/Channel.php [hunk 23]
```diff
@@ -78,8 +81,225 @@ class Channel extends Model
...
+
+            if ($item->nesting) {
+                $channels = $channel->isVisible()->getNested();
+                $iterator = function ($channels) use (&$iterator, &$item, &$theme, $url) {
+                    $branch = [];
+
+                    foreach ($channels as $channel) {
+                        $branchItem = [];
+                        $branchItem['url'] = self::getChannelPageUrl($item->cmsPage, $channel, $theme);
+                        $branchItem['isActive'] = $branchItem['url'] == $url;
```
Comment (bennothommo) [inline models/Channel.php:228]:
```suggestion
                        $branchItem['isActive'] = ($branchItem['url'] === $url);
```

File: models/Channel.php [hunk 24]
```diff
@@ -78,8 +81,225 @@ class Channel extends Model
...
+
+        return $result;
+    }
+
+    /**
+     * Returns URL of a channel page.
+     *
+     * @param $pageCode
+     * @param $channel
+     * @param $theme
```
Comment (bennothommo) [inline models/Channel.php:272]:
```suggestion
     * @param string $pageCode
     * @param \Winter\Forum\Model\Channel $channel
     * @param \Cms\Classes\Theme $theme
```

File: models/Channel.php [hunk 25]
```diff
@@ -78,8 +81,225 @@ class Channel extends Model
...
+
+        return $result;
+    }
+
+    /**
+     * Returns URL of a channel page.
+     *
+     * @param string $pageCode
+     * @param \Winter\Forum\Models\Channel $channel
+     * @param \Cms\Classes\Theme $theme
```
Comment (bennothommo) [inline models/Channel.php:272]:
```suggestion
     * @param string $pageCode
     * @param \Winter\Forum\Models\Channel $channel
     * @param \Cms\Classes\Theme $theme
     * @return string|null
```

File: models/Topic.php [hunk 26]
```diff
@@ -318,4 +321,128 @@ public function lockTopic()
...
+
+        return $result;
+    }
+
+    /**
+     * Returns URL of a topic page.
+     *
+     * @param $pageCode
+     * @param $topic
+     * @param $theme
```
Comment (bennothommo) [inline models/Topic.php:421]:
```suggestion
     * @param string $pageCode
     * @param \Winter\Forum\Models\Topic $topic
     * @param \Cms\Classes\Theme $theme
     * @return string|null
```

**Review by bennothommo — APPROVED** [review]

**Comment by bennothommo** [comment]
@RomainMazB great work on this!

**Pull request merged**